pkg/networkservice/chains/nsmgr/server.go

```diff
@@ -14,65 +14,110 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package nsmgr provides a Network Service Manager (nsmgr), but interface and implementation
+// Package nsmgr provides a Network Service Manager (nsmgrServer), but interface and implementation
 package nsmgr
 
 import (
+	"context"
+	"net/url"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/nsmgr/authz"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/connect"
+	"github.com/networkservicemesh/sdk/pkg/registry/nsmreg"
+
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
 	"github.com/networkservicemesh/api/pkg/api/registry"
-	"google.golang.org/grpc"
 
 	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/client"
-	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/endpoint"
-	"github.com/networkservicemesh/sdk/pkg/networkservice/common/connect"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/authorize"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/common/discover"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/common/localbypass"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/common/roundrobin"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/adapters"
 	adapter_registry "github.com/networkservicemesh/sdk/pkg/registry/core/adapters"
-	chain_registry "github.com/networkservicemesh/sdk/pkg/registry/core/chain"
+	"github.com/networkservicemesh/sdk/pkg/registry/memory"
 	"github.com/networkservicemesh/sdk/pkg/tools/addressof"
+
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/endpoint"
+	"github.com/networkservicemesh/sdk/pkg/tools/callback"
 	"github.com/networkservicemesh/sdk/pkg/tools/token"
 )
 
 // Nsmgr - A simple combintation of the Endpoint, registry.NetworkServiceRegistryServer, and registry.NetworkServiceDiscoveryServer interfaces
 type Nsmgr interface {
 	endpoint.Endpoint
+
 	registry.NetworkServiceRegistryServer
 	registry.NetworkServiceDiscoveryServer
+	registry.NsmRegistryServer
+
+	callback.CallbackServiceServer
 }
 
-type nsmgr struct {
+type nsmgrServer struct {
 	endpoint.Endpoint
+
 	registry.NetworkServiceRegistryServer
 	registry.NetworkServiceDiscoveryServer
+	registry.NsmRegistryServer
+
+	storage memory.Storage
+	callback.CallbackServiceServer
```

Comment:
What happens if I have more than one service that needs to use Callbacks ?

Reply:
Any of callbacks are identified by ID's passed with HandleCallbacks, we could distinguish between types of clients by passing out our identity function and use different ids to callback to different services. So do not see any obstacles here.